amdsharktuner/amdsharktuner/process_utils.py

```diff
@@ -0,0 +1,178 @@
+
+
+class MultiprocessExecutor:
+    """
+    Wrapper of multiprocessing + progress bar + time budget.
+
+    Example:
+        executor = MultiprocessExecutor(num_workers=4)
+        executor.run(task_list, worker_func)
+    """
+
+    def __init__(
+        self,
+        num_workers: int,
+        initializer=None,
+        initializer_inputs=(),
+        time_budget=None,
+    ):
+        self.num_workers = num_workers
+        self.initializer = initializer
+        self.initializer_inputs = initializer_inputs
+        self.time_budget = time_budget
+
+    def run(self, task_list, worker_fn):
+        results = []
+
+        # Create a multiprocessing pool.
+        sigint_handler = signal.signal(signal.SIGINT, signal.SIG_IGN)
+        with multiprocessing.Pool(
+            self.num_workers, self.initializer, self.initializer_inputs
+        ) as worker_pool:
+            signal.signal(signal.SIGINT, sigint_handler)
+            # Use tqdm to create a progress bar.
+            with tqdm(total=len(task_list)) as pbar:
+                try:
+                    # Use imap_unordered to asynchronously execute the worker function on each task.
+                    for result in worker_pool.imap_unordered(worker_fn, task_list):
+                        results.append(result)
+                        pbar.update(1)  # Update progress bar.
+                        # If time limit is reached, stop progress wrapper.
+                        if self.time_budget is not None and self.time_budget.expired():
+                            logging.warning(
+                                f"Time limit reached, total {len(results)} results collected"
+                            )
+                            worker_pool.terminate()
+                            worker_pool.join()
+                            return results
+                except KeyboardInterrupt:
+                    # If Ctrl+C is pressed, terminate all child processes.
+                    worker_pool.terminate()
+                    worker_pool.join()
+                    sys.exit(1)  # Exit the script.
+
+        return results
+
+
+@dataclass
+class RunPack:
+    command: list[str]
+    check: bool = True
+    timeout_seconds: Optional[float] = None
+
+
+@dataclass
+class RunResult:
+    process_res: Optional[subprocess.CompletedProcess]
+    is_timeout: bool
+
+
+def run_command(run_pack: RunPack) -> RunResult:
+    """
+    Wrapper around subprocess.run() with optional timeout and error handling.
+
+    Example:
+        result = process_utils.run_command(
+            process_utils.RunPack(
+                command=["echo", "hello"],
+                check=True,
+            )
+        )
+    """
+    command = run_pack.command
+    check = run_pack.check
+    timeout_seconds = run_pack.timeout_seconds
+
+    result = None
+    is_timeout = False
+    try:
+        # Convert the command list to a command string for logging.
+        command_str = " ".join(command)
```

Comment:
I think there's a library for helper that handles quoting etc: https://docs.python.org/3/library/shlex.html#shlex.join